hil/ext/switches/ovs.py

```diff
@@ -0,0 +1,248 @@
+
+# Copyright 2013-2017 Massachusetts Open Cloud Contributors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the
+# License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS
+# IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied.  See the License for the specific language
+# governing permissions and limitations under the License.
+"""A switch driver for OpenVswitch.
+"""
+
+import re
+import logging
+import schema
+import os
+import ast
+from commands import getstatusoutput
+
+from hil.model import db, Switch, BigIntegerType, SwitchSession
+logger = logging.getLogger(__name__)
+
+
+class VlanError(Exception):
+    """ Raise this exception when vlan cannot be added to the port."""
+
+
+class PortInfoNotFound(Exception):
+    """ Raise this exception when this driver is unable to fetch
+    information for a given port.
+    """
+
+
+class OVS_RequestFailed(Exception):
+    """ Raise this exception when a requestOVS failure
+    does not follow into any of the above exceptions.
+    """
+
+
+class Ovs(Switch, SwitchSession):
+    """ Driver for openvswitch. """
+    api_name = 'http://schema.massopencloud.org/haas/v0/switches/ovs'
+    dir_name = os.path.dirname(__file__)
+    __mapper_args__ = {
+        'polymorphic_identity': api_name,
+    }
+
+    id = db.Column(
+            BigIntegerType, db.ForeignKey('switch.id'), primary_key=True
+            )
+    hostname = db.Column(db.String, nullable=False)
+    username = db.Column(db.String, nullable=False)
+    password = db.Column(db.String, nullable=False)
+
+    @staticmethod
+    def validate(kwargs):
+        schema.Schema({
+            'username': basestring,
+            'hostname': basestring,
+            'password': basestring,
+        }).validate(kwargs)
+
+    def session(self):
+        """ required from superclass. """
+        return self
+
+    def disconnect(self):
+        """ required from superclass. """
+
+    @staticmethod
+    def validate_port_name(port):
+        """ This driver accepts any string as port name."""
+
+    def str2list(self, a_string):
+        """ Converts a string representation of list to list.
+        Empty list is put as None.
+        """
+        if a_string[0] == '[' and a_string[1] == ']':
+            return None
+        else:
+            a_string = a_string.replace("[", "['").replace("]", "']")
+            a_string = a_string.replace(",", "','")
+            a_list = ast.literal_eval(a_string)
+            a_list = [ele.strip() for ele in a_list]
+            return a_list
+
+    def str2dict(self, a_string):
```

Comment:
I don't like the approach to invoking ovs commands here. You're doing a bunch of string munging to marshal the commands through sudo (which is going to be error prone), and it's also just fairly complex. Advice:

* Avoid passing strings through the shell. If you use the `subprocess` module, you can invoke commands directly, so you don't have to escape their arguments.
* Instead of doing your own custom marshalling, just use json.dumps/json.loads.

Another idea: rather than dealing sudo asking for a password either:

* Figure out a way for the user to not need sudo access. I don't know enough about ovs to know what the situation is, but sometimes you just need to be a member of some group, or the bridge needs to be assigned to your user when it's created, or you need to do something with policykit, or... (How do the openstack services work here?)
* Failing that, require that the user is able to run the requisite commands via sudo *without* a password (/etc/sudoers allows you to do this on a command-by-command basis, including restricting arguments)
* Or, if you want to get fancier: write a wrapper command that encapsulates the ovs functionality you actually need, and ask the user to configure sudo to let them run *that* with no password.

Reply:
I will think of the options you presented. I was kind of expecting this.
Since I myself spent quite some time trying various options and opted the one used here for the sake of simplicity. 

other options required, setting up sudo profile which proved to be tedious. 
I will try the subprocess option and try others to see if I can improve the code. 

Thanks for the feedback.